### ISSUE TITLE
Can Not Deref Tags from another Object during creation

### ISSUE THREAD

**Issue by Joe-Kollin** [hashicorp/terraform-cdk#1659]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

Tagging with RouteTables, Routes, and NAT Gateways

### Expected Behavior

`
// Create NAT gateways
    this.natGateways = this.publicSubnets.map((subnet, index) => {
      return new vpc.NatGateway(
        this,
        `${this.config.environment}-${this.config.name}-public-subnet-nat-test-${index}`,
        {
          dependsOn: this.publicSubnets,
          allocationId: this.natGatewayEips[index].allocationId,
          subnetId: subnet.id,
          tags: {
            az: subnet.availabilityZone,
          },
        }
      );
    });

// Create subnets tables
    this.privateRouteTables = this.privateSubnets.map((subnet, index) => {
      return new vpc.RouteTable(
        this,
        `${this.config.environment}-${this.config.name}-private-subnet-route-table-${index}`,
        {
          vpcId: this.vpc.id,
          tags: {
            'name': `${this.config.environment}-${this.config.name}-private-subnet-route-table-${subnet.id}`,
            'network': 'private',
            'subnetId': subnet.id,
            'az': subnet.availabilityZone,
          },
        }
      );
    });

    // Create route table routes for each subnet
    this.privateRouteTableRoutes = this.privateRouteTables.map((routeTable, index) => {
      const natGatewayId = this.natGateways?.find((gateway) => {
        return routeTable.tags.az === gateway.tags.az;
      })?.id;

      if (natGatewayId) return [];

      return new vpc.Route(
        this,
        `${this.config.environment}-${this.config.name}-private-subnet-route-${index}`,
        {
          routeTableId: routeTable.id,
          destinationCidrBlock: '0.0.0.0/0',
          natGatewayId: natGatewayId,
        }
      );
    }).flat();
`

### Actual Behavior

I want to reference tags of a created object and the deref does not work

### Steps to Reproduce

Create a resource with tags and try to reference tag values in the creation of another resource by dereferencing the tags of the former.

- #0000

**Comment by ansgarm** [comment]
Hi @Joe-Kollin 👋
It is currently [not possible to iterate](https://github.com/hashicorp/terraform-cdk/issues/1089) on resources that are only known at runtime. References are resolved to Terraform references, so the actual values behind those will never be available in TypeScript while the code is synthesized.

However, as you seem to be iterating over an array of subnets (`this.publicSubnets`), that probably also are created while the stack is synthesized, you could change your code to contain a single loop (over the AZs) and in there create the resources that are supposed to belong to that AZ.

Another option would be to use e.g. `subnet.availabilityZoneInput` which returns the value that was passed to the Subnet when it was created in your code. This is also available for `.tagsInput` and all other attributes that you set. Those are available at synth time. If you post a full example (e.g. in a separate Github repo) I could create a PR with that change if you want.

**Comment by Joe-Kollin** [comment]
Thanks! No need to fix the changes, I can make them, but that helps

**Comment by skorfmann** [comment]
Perhaps it would make sense to enrich the documentation around this behaviour a bit. Had a quick look at https://www.terraform.io/cdktf/concepts/tokens which could have a section (or even a completely new page) around dynamic behaviour like iterations. What do you think?

**Comment by Joe-Kollin** [comment]
Yeah I think that would help a lot. Using the `.tagsInput` or those input attributes in general has helped me a ton, and a quick spot to find recommendations or best practices for that sort of thing as well as cross stack references would be awesome.

**Comment by jsteinich** [comment]
When we initially added the `xInput` properties it was assumed that we'd eventually remove them. Given the direction that we've gone with tokens / references it seems like they are probably here to stay. Do you agree @skorfmann? If so, we can definitely improve the inline and external documentation around using them.
